src/rules/no-classfield-shadowing.ts

```diff
@@ -0,0 +1,49 @@
+/**
+ * @fileoverview Disallows properties shadowed as class fields
+ * @author Michel Langeveld <https://github.com/michellangeveld>
+ */
+
+import {Rule} from 'eslint';
+import * as ESTree from 'estree';
+import {getClassFields, getPropertyMap} from '../util';
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+const rule: Rule.RuleModule = {
+  meta: {
+    docs: {
+      description: 'Disallows properties shadowed as class field',
+      recommended: true,
+      url: 'https://github.com/43081j/eslint-plugin-lit/blob/master/docs/rules/no-classfield-shadowing.md'
+    },
+    schema: [],
+    messages: {
+      noClassfieldShadowing:
+        'The {{ prop }} property is a class field which has the same as ' +
+        'static property which could have unintended side-effects.'
+    }
+  },
+
+  create(context): Rule.RuleListener {
+    return {
+      ClassDeclaration: (node: ESTree.Class): void => {
```

Comment:
we should probably return early if this isn't a lit element class

you can then add a test for that case:

```ts
class Foo {
  field;
  static properties = {foo: {type: String}};
}
```

in other rules we basically check if the super class is `LitElement`. I'd probably extract that condition into a util for now, and call the util here (something like `isLitClass`).

it does mean deeper inheritance won't be supported yet, but the other rules are already like that so we're probably fine

Reply:
If we go this route, then we should also include mixins, otherwise this rule will probably be useless for us at ING because we usually use _at least_ `ScopedElementsMixin`, but often also other mixins. So we should add a test for something like:

```js
class Foo extends A(B(LitElement)) {
  foo;
  static properties = { foo: {} };
}
```

Reply:
it also suggests a few other rules are currently not running on your ING code (since they look for a `LitElement` super class).

if we extract the logic into a util and add mixin support, those will start working for you too

you could do a simple approach to this without having to resolve symbols etc, just by traversing the expressions

something like:

```ts
function getSuperClass(node: ESTree.Class): ESTree.Identifier|null {
  if (node.superClass.type === 'Identifier') {
    return node.superClass;
  }

  if (node.superClass.type === 'CallExpression') {
    let superClass = node.superClass;
    while (superClass && superClass.type === 'CallExpression') {
      superClass = superClass.arguments.find((arg) => arg.type === 'CallExpression' || arg.type === 'Identifier');
    }

    if (superClass) {
      return superClass;
    }
  }

  return null;
}